libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,363 @@
+  environment: IsographEnvironment,
+  node: OptimisticNode | NetworkResponseNode | StartUpdateNode | null,
+  mutableEncounteredIds: EncounteredIds,
+) {
+  while (node && node?.kind !== 'OptimisticNode') {
+    const data = 'startUpdate' in node ? node.startUpdate().data : node.data;
+    compareData(node.data, data, mutableEncounteredIds);
+    mergeDataLayer(environment.store.data, data);
+    node = node.parentNode;
+  }
+  return node;
+}
+
+function reexecuteUpdates(
+  environment: IsographEnvironment,
+  node: OptimisticNode | NetworkResponseNode | StartUpdateNode | null,
+  mutableEncounteredIds: EncounteredIds,
+) {
+  while (node !== null) {
+    const oldData = node.data;
+    if ('startUpdate' in node) {
+      node.data = node.startUpdate().data;
+    }
+    compareData(oldData, node.data, mutableEncounteredIds);
+    node.childNode = environment.store;
+    environment.store.parentNode = node;
+    environment.store = node;
+
+    node = node.parentNode;
+
+    environment.store.parentNode = null;
+  }
+}
+
+function makeRootNode(environment: IsographEnvironment, node: OptimisticLayer) {
+  node.parentNode = null;
+  environment.store = node;
+}
+
+function replaceOptimisticNodeWithNetworkResponseNode(
+  environment: IsographEnvironment,
+  optimisticNode: OptimisticNode,
+  data: DataLayer,
+  encounteredIds: EncounteredIds,
+) {
+  if (optimisticNode.childNode.kind === 'BaseNode') {
+    mergeDataLayer(optimisticNode.childNode.data, data);
+
+    makeRootNode(environment, optimisticNode.childNode);
+    const node = mergeParentNodes(
+      environment,
+      optimisticNode.parentNode,
+      encounteredIds,
+    );
+    reexecuteUpdates(environment, node, encounteredIds);
+  } else if (optimisticNode.childNode.kind === 'NetworkResponseNode') {
+    mergeDataLayer(optimisticNode.childNode.data, data);
+
+    makeRootNode(environment, optimisticNode.childNode);
+    reexecuteUpdates(environment, optimisticNode.parentNode, encounteredIds);
+  } else if (optimisticNode.parentNode?.kind === 'NetworkResponseNode') {
+    const networkResponseNode = optimisticNode.parentNode;
+    mergeDataLayer(data, networkResponseNode.data);
+    networkResponseNode.data = data;
+
+    networkResponseNode.childNode = optimisticNode.childNode;
+    optimisticNode.childNode.parentNode = networkResponseNode;
+
+    const parentNode = optimisticNode.parentNode.parentNode;
+    makeRootNode(environment, networkResponseNode);
+    reexecuteUpdates(environment, parentNode, encounteredIds);
+  } else {
+    const networkResponseNode: NetworkResponseNode = {
+      kind: 'NetworkResponseNode',
+      data,
+      childNode: optimisticNode.childNode,
+      parentNode: null,
+    };
+
+    makeRootNode(environment, networkResponseNode);
+    reexecuteUpdates(environment, optimisticNode.parentNode, encounteredIds);
+  }
+}
+
+export type OptimisticLayer =
+  | OptimisticNode
+  | NetworkResponseNode
+  | StartUpdateNode
+  | BaseNode;
```

Comment:
This is more complicated than I expected! I didn't expect us to need four layer types. But that is in fact the correct way to model this!